### ISSUE TITLE
Implement bulk deletion

### ISSUE THREAD

**Issue by roji** [dotnet/efcore#27550]
The SQL Server provider implements bulk insertion, i.e. collapsing multiple insertions into a single multi-row MERGE (or INSERT) statement. We could do the same with deletion, so instead of multiple DELETE statements, we'd have:

```sql
DELETE FROM Foo WHERE Id1 IN (1, 2, 3);
```

Benchmark:

|                            Method | DatabaseType | NumRows |       Mean |    Error |    StdDev |     Median | Ratio | RatioSD |
|---------------------------------- |------------- |-------- |-----------:|---------:|----------:|-----------:|------:|--------:|
| Separate_deletes_with_transaction |    SqlServer |       2 | 1,277.5 us |  8.46 us |   7.50 us | 1,276.8 us |  1.00 |    0.00 |
|                  Separate_deletes |    SqlServer |       2 |   559.8 us | 13.75 us |  40.54 us |   572.0 us |  0.44 |    0.02 |
|             Single_delete_with_in |    SqlServer |       2 |   542.0 us | 10.71 us |  18.48 us |   545.5 us |  0.43 |    0.01 |
|                                   |              |         |            |          |           |            |       |         |
| Separate_deletes_with_transaction |    SqlServer |       4 | 1,306.5 us | 25.97 us |  46.83 us | 1,317.7 us |  1.00 |    0.00 |
|                  Separate_deletes |    SqlServer |       4 |   611.7 us | 12.80 us |  37.34 us |   619.8 us |  0.46 |    0.05 |
|             Single_delete_with_in |    SqlServer |       4 |   548.2 us | 13.25 us |  38.85 us |   556.5 us |  0.41 |    0.04 |
|                                   |              |         |            |          |           |            |       |         |
| Separate_deletes_with_transaction |    SqlServer |       6 | 1,350.6 us | 10.11 us |   9.46 us | 1,352.0 us |  1.00 |    0.00 |
|                  Separate_deletes |    SqlServer |       6 |   666.4 us | 13.13 us |  26.23 us |   673.2 us |  0.48 |    0.03 |
|             Single_delete_with_in |    SqlServer |       6 |   559.7 us | 11.12 us |  15.23 us |   561.2 us |  0.41 |    0.01 |
|                                   |              |         |            |          |           |            |       |         |
| Separate_deletes_with_transaction |    SqlServer |      10 | 1,399.1 us | 27.74 us |  81.35 us | 1,414.1 us |  1.00 |    0.00 |
|                  Separate_deletes |    SqlServer |      10 |   768.3 us | 20.72 us |  61.11 us |   785.2 us |  0.55 |    0.07 |
|             Single_delete_with_in |    SqlServer |      10 |   572.7 us |  9.29 us |   8.23 us |   572.2 us |  0.44 |    0.08 |
|                                   |              |         |            |          |           |            |       |         |
| Separate_deletes_with_transaction |    SqlServer |      20 | 1,590.3 us | 14.51 us |  12.87 us | 1,590.3 us |  1.00 |    0.00 |
|                  Separate_deletes |    SqlServer |      20 | 1,049.5 us | 35.67 us | 105.17 us | 1,084.4 us |  0.65 |    0.06 |
|             Single_delete_with_in |    SqlServer |      20 |   598.0 us | 11.56 us |  17.65 us |   602.5 us |  0.37 |    0.01 |

Notes/caveats:

* This allows removing the transaction if the bulk delete is the only thing being saved (single command).
  * A big part of the perf gap from the transaction would go away if we do #27532
  * However, even without a transaction, the gap between multiple commands and a single one is significant, making this a worthy optimization.
* For composite keys, row values can be used on MySQL and PostgreSQL:
  * `DELETE FROM Foo WHERE (Id1, Id2) IN ((1, 2), (3, 4))`
  * This isn't supported by SQL Server or SQLite (they could opt out of this optimization for composite keys).
* At least in theory, multiple (identical) DELETEs could reuse the same query plan, where single multi-row DELETEs for varying row numbers would not.

<details>
<summary>Benchmark code</summary>

```c#
BenchmarkRunner.Run<Benchmark>();

public class Benchmark
{
    [Params(DatabaseType.Postgresql, DatabaseType.SqlServer)]
    public DatabaseType DatabaseType { get; set; }

    [Params(2, 4, 6, 10, 20)]
    public int NumRows { get; set; }

    private DbConnection _connection;
    private DbCommand _command;

    private async Task Setup()
    {
        _connection = DatabaseType switch
        {
            DatabaseType.Postgresql => new NpgsqlConnection("Host=localhost;Username=test;Password=test"),
            DatabaseType.SqlServer => new SqlConnection("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Trust Server Certificate=true"),
            _ => throw new ArgumentOutOfRangeException()
        };

        await _connection.OpenAsync();

        await using var command = _connection.CreateCommand();
        command.CommandText = "DROP TABLE IF EXISTS foo; CREATE TABLE foo (id INT, data INT)";
        await command.ExecuteNonQueryAsync();
    }

    [GlobalSetup(Targets = new[] { nameof(Separate_deletes), nameof(Separate_deletes_with_transaction) })]
    public async Task Setup_Separate_deletes()
    {
        await Setup();

        _command = _connection.CreateCommand();
        _command.CommandText = new StringBuilder()
            .AppendJoin(" ", Enumerable.Range(0, NumRows).Select(i => $"DELETE FROM foo WHERE id = @p{i};"))
            .ToString();

        for (var i = 0; i < NumRows; i++)
        {
            var param = _command.CreateParameter();
            param.ParameterName = "p" + i;
            param.Value = i;
            _command.Parameters.Add(param);
        }
    }

    [GlobalSetup(Target = nameof(Single_delete_with_in))]
    public async Task Setup_Single_delete_with_in()
    {
        await Setup();

        _command = _connection.CreateCommand();
        _command.CommandText = new StringBuilder()
            .Append("DELETE FROM foo WHERE id IN (")
            .AppendJoin(", ", Enumerable.Range(0, NumRows).Select(i => "@p" + i))
            .Append(")")
            .ToString();

        for (var i = 0; i < NumRows; i++)
        {
            var param = _command.CreateParameter();
            param.ParameterName = "p" + i;
            param.Value = i;
            _command.Parameters.Add(param);
        }
    }

    [Benchmark(Baseline = true)]
    public async Task Separate_deletes_with_transaction()
    {
        var tx = _connection.BeginTransaction();
        _command.Transaction = tx;
        await _command.ExecuteNonQueryAsync();
        tx.Commit();
    }

    [Benchmark]
    public Task Separate_deletes()
        => _command.ExecuteNonQueryAsync();

    [Benchmark]
    public Task Single_delete_with_in()
        => _command.ExecuteNonQueryAsync();
}

public enum DatabaseType
{
    Postgresql,
    SqlServer
}

```

</details>

**Comment by roji** [comment]
Note: the problem here is concurrency detection, see #27562.

**Comment by roji** [comment]
If we do decide to keep concurrency checks, we could do a single query after multiple updates to get the number of rows with the IDs we wanted to delete (and expect zero to come back). However, if it's possible for keys to repeat (i.e. not IDENTITY/GUID), then it's possible for a concurrent insert to add the row we just deleted, causing a spurious concurrency exception. Note that this is a problem even if we're in a transaction, since newly-committed rows do appear from outside the transaction; the only way to protect against this AFAIK is via the serializable isolation level, which we definitely shouldn't do.

Bottom line, unless we decide to get rid of the concurrency checks, we probably shouldn't do anything here.

**Comment by roji** [comment]
Thinking about it some more, it seems the the concurrency check isn't actually a problem here. One can do:

```sql
DELETE FROM Foo WHERE id IN (1, 2, 3) RETURNING 1;
```

This returns a row with 1 for each row deleted. This allows us to detect when a concurrency issue, but does not allow us to know exactly which entry was the problematic one.

**Comment by roji** [comment]
Design discussion:

* We can preserve full concurrency checking by returning the IDs of the deleted rows, instead of 1 (thanks @bricelam). Probably requires some client-side changes.
* Re query plan fragmentation, we already have this problem with inserts, and one mitigation is the limitation of the batch size. No need to do anything specific here.
* As the optimization isn't huge, we'll see whether we implement it for 7.0.

**Comment by roji** [comment]
Note to self: when implementing this, note that a single bulk delete statement is *not* safe for running without transactions, as long as concurrency checking is enabled - for the same reasons as in https://github.com/dotnet/efcore/issues/27532#issuecomment-1073059572. In other words, if less rows were deleted than expected, we want to throw and roll back, but if there's no transaction the change has already been committed and there's nothing to roll back

**Comment by roji** [comment]
As an alternative, if in the update pipeline we can figure out that we're in a client cascade sort of scenarios ("delete all Posts who's Blog has Id 1"), we can generate a single, simple `DELETE FROM Posts p WHERE p.BlogId = foo` (see https://github.com/dotnet/efcore/issues/35831#issuecomment-2748830692).
